datastore/providers/mongodb_atlas_datastore.py

```diff
@@ -0,0 +1,252 @@
+
+
+class MongoDBAtlasDataStore(DataStore):
+
+    def __init__(
+        self,
+        atlas_connection_uri: str = MONGODB_CONNECTION_URI,
+        index_name: str = MONGODB_INDEX,
+        database_name: str = MONGODB_DATABASE,
+        collection_name: str = MONGODB_COLLECTION,
+        oversampling_factor: float = OVERSAMPLING_FACTOR,
+    ):
+        """
+        Initialize a MongoDBAtlasDataStore instance.
+
+        Parameters:
+        - index_name (str, optional): Vector search index. If not provided, default index name is used.
+        - database_name (str, optional): Database. If not provided, default database name is used.
+        - collection_name (str, optional): Collection. If not provided, default collection name is used.
+        - oversampling_factor (float, optional): Oversampling factor for data augmentation.
+                                                 Default is OVERSAMPLING_FACTOR.
+
+        Raises:
+        - ValueError: If index_name is not a valid string.
+
+        Attributes:
+        - index_name (str): Name of the index.
+        - database_name (str): Name of the database.
+        - collection_name (str): Name of the collection.
+        - oversampling_factor (float): Oversampling factor for data augmentation.
+        """
+
+        self.atlas_connection_uri = atlas_connection_uri
+        self.oversampling_factor = oversampling_factor
+        self.database_name = database_name
+        self.collection_name = collection_name
+
+        if not (index_name and isinstance(index_name, str)):
+            raise ValueError("Provide a valid index name")
+        self.index_name = index_name
+
+        # TODO: Create index via driver https://jira.mongodb.org/browse/PYTHON-4175
+        # self._create_search_index(num_dimensions=1536, path="embedding", similarity="dotProduct", type="vector")
+
+    @cached_property
+    def client(self):
+        return self._connect_to_mongodb_atlas(
+            atlas_connection_uri=MONGODB_CONNECTION_URI
+        )
+
+    async def upsert(
+        self, documents: List[Document], chunk_token_size: Optional[int] = None
+    ) -> List[str]:
+        """
+        Takes in a list of Documents, chunks them, and upserts the chunks into the database.
+        Return a list the ids of the document chunks.
+        """
+        chunks = get_document_chunks(documents, chunk_token_size)
+        return await self._upsert(chunks)
+
+    async def _upsert(self, chunks: Dict[str, List[DocumentChunk]]) -> List[str]:
+        """
+        Takes in a list of document chunks and inserts them into the database.
+        Return a list of document ids.
+        """
+        documents_to_upsert = []
+        inserted_ids = []
+        for chunk_list in chunks.values():
+            for chunk in chunk_list:
+                inserted_ids.append(chunk.id)
+                documents_to_upsert.append(
+                        UpdateOne({'_id': chunk.id}, {"$set": chunk.dict()}, upsert=True)
```

Comment:
Correct me if I am wrong but I have something to point out:
1. The upsert will return the chunk IDs instead of the document IDs. So If I insert N documents, I will get more (or equal if they are tiny) than N IDs. So the docstring isn't right.
2. if chunk is a basemodel from pydantic use model_dump(), I think dict is deprecated?

I do like the idea, it is a better approach without a doubt, I think is better to return the documents ids, if the document got splited in the same way the ids will be the same and everything will work.

Reply:
The other thing is, what is going to happen if a document with 5 chunks is inserted, then modified, but the new document is 2 chunks shorter? I think the old 2 chunks are still existing. So, if a chapter of a document is removed, a query that matches the old part of the document can cause a false match.

Reply:
@WaVEV RE: return id's. I changed it to be the chunk ids because this is all that the datastore knows about. Itis a simple change to make if need be. As far as I can tell, the ids returned are not used. If there are conventions, we are glad to follow them.

RE: `chunk.dict()` I inferred the usage from a previous commit of yours which had done similar: `"metadata": document_chunk.metadata.dict()`. This removes the helper function that you had. MongoDB can consume the data as-is.